### PR TITLE
[DOCS-1589][DOCS-1590] Docs/manager/konnect plugin examples

### PR DESCRIPTION
**Pull request by HCloward**
### Review
@lena-larionova 

### Summary
Updated plugin layout template to include tabs for Konnect and Kong Manager when adding plugins via services, routes, and globally.

### Reason
[DOCS-1589](https://konghq.atlassian.net/browse/DOCS-1589)
[DOCS-1590](https://konghq.atlassian.net/browse/DOCS-1590)

### Testing
[Example with service, route, consumer, and global options](https://deploy-preview-2685--kongdocs.netlify.app/hub/kong-inc/rate-limiting-advanced/)
[Example with no required parameters](https://deploy-preview-2685--kongdocs.netlify.app/hub/kong-inc/jwt/)
[Example with checkbox options](https://deploy-preview-2685--kongdocs.netlify.app/hub/kong-inc/application-registration/)

### CODE REVIEW

File: app/_layouts/extension.html [hunk 1]
```diff
@@ -105,18 +132,51 @@
...
+
+  `<service>` is the `id` or `name` of the Service that this plugin
+    configuration will target.
 
 {% endnavtab %}
+{% unless page.params.manager_examples == false %}
+{% navtab Kong Manager %}
+Configure this plugin on a Service:
+
+1. In Kong Manager, select the workspace.
```
Comment (tlblessing) [inline app/_layouts/extension.html:154]: Is there any way to  shorten and combine the nav steps?

File: app/_layouts/extension.html [hunk 2]
```diff
@@ -105,18 +132,51 @@
...
+
+6. If the option is available, set the plugin scope to **Scoped**.
+
+7. Add the Service field with the Service name and ID if it 
+is not already prefilled. 
+
+{% if config_required_fields_konnect == empty %}
+8. Click **Create** to save.
+  {% else %}  
+8. Enter the following required [parameters](#parameters), updating the default/sample values as needed:
```
Comment (tlblessing) [inline app/_layouts/extension.html:173]: should avoid slash constructions default/sample

File: app/_layouts/extension.html [hunk 3]
```diff
@@ -178,19 +241,44 @@
...
+    <strong>Note:</strong> If the plugin is greyed out, then it is not available
+    for your product tier. See <a href="">Kong Gateway tiers</a>.
+    </div>
+
+  {% if config_required_fields_konnect == empty %}
+5. Click **Create** to save.
+  {% else %}  
+5. Enter the following required [parameters](#parameters), updating the default/sample values as needed:
+    {{ config_required_fields_konnect }}
+6. Click **Create** to save.
```
Comment (tlblessing) [inline app/_layouts/extension.html:269]: "Click Create to save" sounds like there is an issue with GUI labels. Don't need the "to save" because it is actually creating the plugin, not saving an existing config.

```suggestion
6. Click **Create**.
```

File: app/_layouts/extension.html [hunk 4]
```diff
@@ -178,19 +241,44 @@
...
 {% endunless %}
 {% endnavtabs %}
 {% endif %}
 
-`<service>` is the `id` or `name` of the Service that this plugin
-  configuration will target.
-
 {% if page.params.api_id %}
-  > **Note:** The legacy API entity has been **deprecated** in favor of Services
+<div class="alert alert-ee blue"><strong>Note:</strong>The legacy API entity has been **deprecated** in favor of Services
```
Comment (tlblessing) [inline app/_layouts/extension.html:277]: Do we even want this here? Murillo has been requesting its removal in my plugin PRs. EE 0.32 is EOL version. Not sure customers know what "EE" is.

File: app/_layouts/extension.html [hunk 5]
```diff
@@ -216,13 +304,47 @@
 
 {{ plugin_config_for_route_admin_api }}
 
+`<route>` is the `id` or `name` of the Route that this plugin configuration
```
Comment (tlblessing) [inline app/_layouts/extension.html:307]: I think per a better style, we should not start with a code element.

```suggestion
The `<route>` is the `id` or `name` of the Route that this plugin configuration
```

File: app/_layouts/extension.html [hunk 6]
```diff
@@ -216,13 +304,47 @@
...
+  will target.
+
 {% else %}
 {% navtabs %}
 {% navtab Admin API %}
 
 {{ plugin_config_for_route_admin_api }}
 
+`<route>` is the `id` or `name` of the Route that this plugin configuration
```
Comment (tlblessing) [inline app/_layouts/extension.html:316]:
```suggestion
The `<route>` is the `id` or `name` of the Route that this plugin configuration
```

File: app/_layouts/extension.html [hunk 7]
```diff
@@ -216,13 +304,47 @@
...
+  will target.
+
 {% endnavtab %}
+{% unless page.params.manager_examples == false %}
+{% navtab Kong Manager %}
+Configure this plugin on a Route:
+
+1. In Kong Manager, select the workspace.
```
Comment (tlblessing) [inline app/_layouts/extension.html:324]: 4 steps for nav seems a bit much. Either something is off with the nav design, or the structure of the content. Suggest combining steps.

File: app/_layouts/extension.html [hunk 8]
```diff
@@ -216,13 +304,47 @@
...
+2. From the Dashboard, select **Routes** in the left navigation.
```
Comment (tlblessing) [inline app/_layouts/extension.html:325]: Should avoid specific references to location (since it can be subject to change), such as "left navigation". Something like "From the Dashboard menu, click Routes and click View for the Route row."

File: app/_layouts/extension.html [hunk 9]
```diff
@@ -216,13 +304,47 @@
...
+3. Click **View** for the Route row.
+4. Scroll down to plugins and click **Add Plugin**.
+5. Find and select the **{{ page.name }}** plugin.
+
+    <div class="alert alert-ee blue">
+    <strong>Note:</strong> If the plugin is greyed out, then it is not available
+    for your product tier. See <a href="">Kong Gateway tiers</a>.
```
Comment (tlblessing) [inline app/_layouts/extension.html:332]: This could also have nothing to do with product tier and more to do with security precautions. For Konnect, tranforms etc are not allowed.

File: app/_layouts/extension.html [hunk 10]
```diff
@@ -216,13 +304,47 @@
...
+    </div>
+
+6. If the option is available, set the plugin scope to **Scoped**.
+
+7. Add the Route ID if it is not already prefilled. 
+
+{% if config_required_fields_konnect == empty %}
+8. Click **Create** to save.
```
Comment (tlblessing) [inline app/_layouts/extension.html:340]:
```suggestion
8. Click **Create**.
```

less is more

File: app/_layouts/extension.html [hunk 11]
```diff
@@ -216,13 +304,47 @@
...
+  {% else %}  
+8. Enter the following required [parameters](#parameters), updating the default/sample values as needed:
```
Comment (tlblessing) [inline app/_layouts/extension.html:342]: smae comments this section as above

File: app/_layouts/extension.html [hunk 12]
```diff
@@ -216,13 +304,47 @@
...
+3. Click **View** for the Route row.
+4. Scroll down to plugins and click **Add Plugin**.
+5. Find and select the **{{ page.name }}** plugin.
+
+    <div class="alert alert-ee blue">
+    <strong>Note:</strong> If the plugin is greyed out, then it is not available
+    for your product tier. See <a href="">Kong Gateway tiers</a>.
+    </div>
+
+6. If the option is available, set the plugin scope to **Scoped**.
```
Comment (tlblessing) [inline app/_layouts/extension.html:335]: Any more information about scoped?

File: app/_layouts/extension.html [hunk 13]
```diff
@@ -283,14 +408,37 @@
...
   will target.
+{% endnavtab %}
+{% endunless %}
+{% unless page.params.konnect_examples == false %}
+{% navtab Konnect %}
+Configure this plugin on a Route:
+
+1. In Konnect, select the Service from the ServiceHub page.
+2. Scroll down to **Versions** and select the version.
+3. Select the route.
```
Comment (tlblessing) [inline app/_layouts/extension.html:422]: inconsistent capitalization of "route". Shane had mentioned that Kong docs have too much initial caps that are "reader speed bumps". Should we even capitalize Service and Route anymore? Unless it is reflecting a specific UI element that is initial-capped?

File: app/_layouts/extension.html [hunk 14]
```diff
@@ -283,14 +408,37 @@
...
+4. Scroll down to **Plugins** and click **Add Plugin**.
+5. Find and select the **{{ page.name }}** plugin.
+
+    <div class="alert alert-ee blue">
+    <strong>Note:</strong> If the plugin is greyed out, then it is not available
+    for your product tier. See <a href="">Kong Gateway tiers</a>.
+    </div>
 
+  {% if config_required_fields_konnect == empty %}
+6. Click **Create** to save.
```
Comment (tlblessing) [inline app/_layouts/extension.html:432]: same comments as above

**Review by tlblessing — CHANGES_REQUESTED** [review]
suggested edits to consider. cool plugin template update.

File: app/_layouts/extension.html (hunk 1 above)
Comment (HCloward) [reply app/_layouts/extension.html:154]: I did for Services and Consumers for Konnect, but for the others I cannot remove any of the steps as they are necessary.

File: app/_layouts/extension.html (hunk 2 above)
Comment (HCloward) [reply app/_layouts/extension.html]: k will replace with or

File: app/_layouts/extension.html (hunk 4 above)
Comment (HCloward) [reply app/_layouts/extension.html]: I have removed it.

File: app/_layouts/extension.html (hunk 5 above)
Comment (HCloward) [reply app/_layouts/extension.html]: This is referencing an item in the code sample and I do not think adding an article is appropriate here.

File: app/_layouts/extension.html (hunk 11 above)
Comment (HCloward) [reply app/_layouts/extension.html]: add or instead

File: app/_layouts/extension.html (hunk 12 above)
Comment (HCloward) [reply app/_layouts/extension.html:324]: not sure what you mean here?

File: app/_layouts/extension.html (hunk 14 above)
Comment (HCloward) [reply app/_layouts/extension.html]: removed "to save" from all instances

**Comment by lena-larionova** [comment]
@HCloward Might want to think about rearranging the tabs to put Manager and Konnect up front. Though now that I think about it, that might end up stretching the page way too much, if everything defaulted to a UI tab first...

**Comment by HCloward** [comment]
> @HCloward Might want to think about rearranging the tabs to put Manager and Konnect up front as "first-class citizens". Though now that I think about it, that might end up stretching the page way too much, if everything defaulted to a UI tab first...

Konnect tends to be a bit shorter than Kong Manager if I put it first page may not be too long. I will make the edit and have you check it out.

File: app/_layouts/extension.html (hunk 12 above)
Comment (lena-larionova) [reply app/_layouts/extension.html:339]: The definition is in the UI, I don't think we need to duplicate the info here. UI currently says:
```
Scoped
Specific Services and/or Routes in this workspace
```

File: app/_layouts/extension.html [hunk 15]
```diff
@@ -210,19 +287,53 @@
...
+3. Click **View** for the route row.
+4. Scroll down to plugins and click **Add Plugin**.
+5. Find and select the **{{ page.name }}** plugin.
+
+    <div class="alert alert-ee blue">
+    <strong>Note:</strong> If the plugin is greyed out, then it is not available
+    for your product tier. See <a href="">Kong Gateway tiers</a>.
+    </div>
+
+6. If the option is available, set the plugin scope to **Scoped**.
```
Comment (lena-larionova) [inline app/_layouts/extension.html:339]:
```suggestion
6. If the option is available, set the plugin to **Scoped**.
```
Maybe shorten this to avoid repetition? It's not explicitly labelled as a plugin scope section anyway.

File: app/_layouts/extension.html [hunk 16]
```diff
@@ -103,20 +130,51 @@
...
+Configure this plugin on a service:
+
+1. In Kong Manager, select the workspace.
+2. From the Dashboard, scroll down to **Services** and click **View** for the service row.
+3. Scroll down to plugins and click **Add Plugin**.
+4. Find and select the **{{ page.name }}** plugin.
+
+    <div class="alert alert-ee blue">
+    <strong>Note:</strong> If the plugin is greyed out, then it is not available
+    for your product tier. See <a href="">Kong Gateway tiers</a>.
```
Comment (lena-larionova) [inline app/_layouts/extension.html:175]:
```suggestion
    for your product tier. See <a href="/enterprise/latest/deployment/licensing/">Kong Gateway tiers</a>.
```
Applies to the rest as well.

File: app/_layouts/extension.html [hunk 17]
```diff
@@ -178,21 +239,37 @@
...
+
+1. In Konnect, select the service on the ServiceHub page.
+2. Scroll down to **Versions** and select the version.
+3. Scroll down to **Plugins** and click **New Plugin**.
+4. Find and select the **{{ page.name }}** plugin.
+
+    <div class="alert alert-ee blue">
+    <strong>Note:</strong> If the plugin is greyed out, then it is not available
+    for your product tier. See <a href="">Kong Gateway tiers</a>.
+    </div>
```
Comment (lena-larionova) [inline app/_layouts/extension.html]: Not sure that you need this note in the Konnect tab at all. I've checked through all of my accounts - free, plus, trial - and the plugins that aren't available just don't appear. That might change in the future, but for now they're just missing.

File: app/_layouts/extension.html [hunk 18]
```diff
@@ -178,21 +239,37 @@
...
-{% if page.params.api_id %}
-  > **Note:** The legacy API entity has been **deprecated** in favor of Services
-  since <a href="https://github.com/Kong/kong/blob/master/CHANGELOG.md#0130---20180322">
-  CE 0.13.0</a> and
-  <a href="https://docs.konghq.com/enterprise/changelog/#0-32">EE 0.32</a>.
-{% endif %}
+{% endnavtab %}
+{% endunless %}
+{% unless page.params.konnect_examples == false %}
+{% navtab Konnect %}
```
Comment (lena-larionova) [inline app/_layouts/extension.html]: Should probably have "Konnect SaaS" as the tab title, even though it'll stretch the tab. 
Maybe shorten the "Kong Manager" tab title to just "Manager" to make up for it.

File: app/_layouts/extension.html [hunk 19]
```diff
@@ -178,21 +239,37 @@
...
+Configure this plugin on a service:
+
+1. In Konnect, select the service on the ServiceHub page.
```
Comment (lena-larionova) [inline app/_layouts/extension.html]:
```suggestion
1. In {{site.konnect_saas}}, select the service on the ServiceHub page.
```
Applies throughout, just not in tab titles - that won't get parsed.

**Comment by tlblessing** [comment]
> @HCloward Might want to think about rearranging the tabs to put Manager and Konnect up front as "first-class citizens". Though now that I think about it, that might end up stretching the page way too much, if everything defaulted to a UI tab first...

@lena-larionova and @HCloward I think the Admin API should be the first tab. The UI should be the last tabs. IMO. 

cc: @GayleNeumann

**Review by lena-larionova — CHANGES_REQUESTED** [review]
Some comments on style/sectioning/etc.

Looks like you accidentally committed your package-lock.json as well? Maybe just do a separate PR to push it so that doesn't keep happening. It's different because you had to update some components, right? So it should actually help.

**Comment by lena-larionova** [comment]
> @lena-larionova and @HCloward I think the Admin API should be the first tab. The UI should be the last tabs. IMO. Especially since the remainder of the examples in the non-generated lower area are in cURL (and some in httpie). This is incongruous content structure.

I brought the idea up because that was part of the original ask from Shane, before Konnect even published: as people get access to the free Gateway version, they increasingly have access to the GUIs, so he wanted that info to be up front and apparent. Of course, the final decision is up to us, as long as we can make a solid argument for any decision...

File: app/_layouts/extension.html [hunk 20]
```diff
@@ -103,20 +130,51 @@
...
+
+5. If the option is available, set the plugin scope to **Scoped**.
+
+6. Add the service name and ID to the **Service** field if it
+is not already prefilled.
 
+{% if config_required_fields_konnect == empty %}
+8. Click **Create**.
+  {% else %}
+8. Enter the following required [parameters](#parameters), updating the default or sample values as needed:
```
[the remark this answers is not in the document]
Comment (HCloward) [reply app/_layouts/extension.html:186]: hmm, well according to the way my conditional statements are written, I am only showing required fields.

File: app/_layouts/extension.html (hunk 7 above)
Comment (HCloward) [reply app/_layouts/extension.html:328]: All steps are required to get to where the user needs to be and each step is on a different page in the UI. I have combined where the steps are on the same page, but I do not think it is a good idea to combine steps that occur on different pages.

File: app/_layouts/extension.html (hunk 8 above)
Comment (HCloward) [reply app/_layouts/extension.html:329]: The location is subject to change, but I need a way to distinguish the left navigation from the top navigation since there are two options. The page is Dashboard, but the navigation remains for all pages, so it is not specific to the Dashboard - however for now it stays on the left side of the screen.

File: app/_layouts/extension.html (hunk 15 above)
Comment (HCloward) [reply app/_layouts/extension.html:339]: k, will fix

**Comment by HCloward** [comment]
> Some comments on style/sectioning/etc.
> 
> Looks like you accidentally committed your package-lock.json as well? Maybe just do a separate PR to push it so that doesn't keep happening. It's different because you had to update some components, right? So it should actually help.

ah, yeah will fix

File: app/_layouts/extension.html (hunk 19 above)
Comment (HCloward) [reply app/_layouts/extension.html]: ah, yeah will fix

File: app/_layouts/extension.html (hunk 18 above)
Comment (HCloward) [reply app/_layouts/extension.html]: yep, will fix

File: app/_layouts/extension.html (hunk 17 above)
Comment (HCloward) [reply app/_layouts/extension.html]: will remove

**Comment by tlblessing** [comment]
@HCloward and @lena-larionova another thing to consider:  is there some way we could consolidate configuring on a service, route, and consumer? I know that not all are applicable to a plugin, but here is so much scrolling going on. The only difference is the placeholder.

**Comment by lena-larionova** [comment]
> @HCloward and @lena-larionova another thing to consider: is there some way we could consolidate configuring on a service, route, and consumer? I know that not all are applicable to a plugin, but here is so much scrolling going on. The only difference is the placeholder.

I agree - I mentioned this to Heather when we were first talking about adding these GUI examples, but I'm not sure what the solution is at the moment. I'd maybe do the work in a subsequent ticket+PR? There's already a lot happening in this one.

**Comment by tlblessing** [comment]
> > @HCloward and @lena-larionova another thing to consider: is there some way we could consolidate configuring on a service, route, and consumer? I know that not all are applicable to a plugin, but here is so much scrolling going on. The only difference is the placeholder.
> 
> I agree - I mentioned this to Heather when we were first talking about adding these GUI examples, but I'm not sure what the solution is at the moment. I'd maybe do the work in a subsequent ticket+PR? There's already a lot happening in this one.

**Comment by tlblessing** [comment]
Didn't mean to close this, was just trying to add a comment. Clicked wrong button by accident. My apologies. 

@HCloward so the variables won't work in the generated areas of the plugin template. So you'll have to spell it out unless @lena-larionova finds a workaround.

I'm not sure when this PR was started, but you might want to rebase your PR to the latest master (Lena, I rebased the old cookie PR and Gayle's PRs so it usually works...) to avoid painful merge conflicts. There have been many changes and merged PRs to plugin hub docs the past few weeks.

Another thing to consider is that some fields that are required have defaults. Most will want to use those defaults. It might be better to inform users or stress to them which fields don't have defaults that they must populate. The way the content is now, it looks like it's more work than it really is to configure plugins. Same holds true for Admin API examples. That is why I think the freeform area is especially useful for documentation, as you can show minimal configs. Some of these freeform areas show with and without a database tabs and so forth so it is not consistent with the autogenerated sections. Just some things we should consider in the future.

**Comment by HCloward** [comment]
> Didn't mean to close this, was just trying to add a comment. Clicked wrong button by accident. My apologies.
> 
> @HCloward so the variables won't work in the generated areas of the plugin template. So you'll have to spell it out unless @lena-larionova finds a workaround.
> 
> Heather, so you list JWT as a plugin with no required parameters, but that is not the case if you look at its schema: https://github.com/Kong/kong-ee/blob/master/kong/plugins/jwt/schema.lua. They are mostly all required (docs are wrong in many spots) but have defaults. This is one of about 20 plugins that still need to have datatypes added and their schemas audited against the plugin hub docs. You can consult my plugin inventory spreadsheet for links to the main plugin schemas. https://docs.google.com/spreadsheets/d/1nZJ0cPkOaZqTGTEICBrTa3TK2xIJqYkclZbJM04eROc/edit#gid=0
> 
> I'm not sure when this PR's branch was started, but you might want to rebase your PR to the latest master (Lena, I rebased the old cookie PR and Gayle's PRs so it usually works...) to avoid potential painful merge conflicts, and to have the most current plugin hub docs. There have been many changes and merged PRs to plugin hub docs the past few weeks.
> 
> Another thing to consider is that some fields that are required have defaults. Most will want to use those defaults. It might be better to inform users or stress to them which fields don't have defaults that they must populate. The way the content is now, it looks like it's more work than it really is to configure plugins. Same holds true for Admin API examples. That is why I think the freeform area is especially useful for documentation, as you can show minimal configs. Some of these freeform areas show with and without a database tabs and so forth so it is not consistent with the autogenerated sections. Just some things we should consider in the future.

I will remove the variable Lena recommended I use as it is not working. I tried rebasing my branch and it wiped out all of my work, so not real keen to do that again. Since this template is simply reading from and printing what is in the docs, if the current information is wrong, not sure how I can fix that unless I fix all of the plugins docs. I am not sure what the freeform area is or how to include it. Perhaps I should just close this PR and let @tlblessing or @lena-larionova do this. It seems I am over my head.

**Comment by lena-larionova** [comment]
@HCloward I can take this back over. There's a lot of good discussion going on here, but you're right in that a lot of it doesn't belong here. Which values the examples pick up depends on individual plugin config settings, and shouldn't be covered here. We need to set a cutoff point and say that this work adds examples for manager and Konnect; any further work should be broken out into other tickets.

@tlblessing I'm going to hold off adding datatypes to the example parser until that's implemented across all plugins. Because when that happens, we'll need to rewrite a good chunk of code, and I don't want to deal with having multiple different parsers for one thing.
This PR has no merge conflicts because it doesn't touch specific plugin files, by the way. The layout is separate from the docs themselves.

**Review by tlblessing — APPROVED** [review]
Thanks for all this awesome work @HCloward  and @lena-larionova !
